### ISSUE TITLE
[mongodb-sharded] connection drops and performance issue

### ISSUE THREAD

**Issue by maxisam** [bitnami/charts#4705]
**Which chart**:

mongodb-sharded

**Describe the bug**

I don't know if this is the bug for this chart but 

If I create a lot of connections like 7000 ish

Mongos starts to fail.

```
":{"$date":"2020-12-11T21:02:07.433+00:00"},"s":"W",  "c":"EXECUTOR", "id":22993,   "ctx":"conn20145","msg":"Terminating session due to error","attr":{"error":{"code":1,"codeName":"InternalError","errmsg":"failed to create service entry worker thread"}}}{"t":{"$date":"2020-12-11T21:02:07.433+00:00"},"s":"I",  "c":"NETWORK",  "id":22944,   "ctx":"listener","msg":"Connection ended","attr":{"remote":"10.42.80.0:47203","connectionId":20145,"connectionCount":7005}}{"t":{"$date":"2020-12-11T21:02:07.434+00:00"},"s":"I",  "c":"NETWORK",  "id":22943,   "ctx":"listener","msg":"Connection accepted","attr":{"remote":"10.42.80.0:10301","connectionId":20146,"connectionCount":7006}}{"t":{"$date":"2020-12-11T21:02:07.434+00:00"},"s":"I",  "c":"-",        "id":22948,   "ctx":"listener","msg":"pthread_create failed","attr":{"error":"Resource temporarily unavailable"}}
```

I can't figure out why. 

I added this to initContainers of mongos but it doesn't make any difference

```
    - name: "adjust-max-map-count"
      image: "busybox:latest"
      securityContext:
        privileged: true
      command: ["sysctl", "-w", "vm.max_map_count=262144"]
```

**To Reproduce**

Start a mongo cluster with 4 shards/3 replica set

use jmeter to connect to mongodb like below and set the thread count like 70 and run 100 times (so it creates 7000 connections)

https://www.blazemeter.com/blog/mongodb-performance-testing-with-jmeter

**Comment by javsalgar** [comment]
Hi,

According to https://jira.mongodb.org/browse/SERVER-31344

```

Thank you for the report. mongod requires ulimits to be appropriately set for its workload. Please increase your settings to our recommended values and reevaluate the performance of your replica set to confirm that the issue has been resolved.

```

Did you try increasing your system ulimits? You may need to tweak the Kubernetes nodes in order to make it work in your containers, but that would depend on the Kubernetes cluster.

**Comment by maxisam** [comment]
Well, I haven't changed it but that was my first thing I checked too. I wasn't sure should I do it in the pod or node. 
But 7000ish is way lower than any default value. And I didn't set any resource limitation.

**Comment by javsalgar** [comment]
Hi,

As we're talking about containers, you can do it in the node or using a privileged container, that would depend on what the platform restrictions. Which are your current ulimit settings?

**Comment by maxisam** [comment]
Here they are. It is the default setting. I think they should be able to handle 7000ish connection.
```
core file size          (blocks, -c) unlimited
data seg size           (kbytes, -d) unlimited
scheduling priority             (-e) 0
file size               (blocks, -f) unlimited
pending signals                 (-i) 23439
max locked memory       (kbytes, -l) 64
max memory size         (kbytes, -m) unlimited
open files                      (-n) 1048576
pipe size            (512 bytes, -p) 8
POSIX message queues     (bytes, -q) 819200
real-time priority              (-r) 0
stack size              (kbytes, -s) 8192
cpu time               (seconds, -t) unlimited
max user processes              (-u) unlimited
virtual memory          (kbytes, -v) unlimited
file locks                      (-x) unlimited
```

**Comment by javsalgar** [comment]
Hi,

Thanks for the information. I think it's also important to know the number of Kubernetes nodes you have in your cluster. Do you have any other applications running? How are the mongo instances scheduled in the nodes? In any case, I believe this is something that you should also check with the Mongo devs so they confirm that your ulimit settings are correct. At the moment, it is not clear to me that the issue is in the chart itself.

**Comment by maxisam** [comment]
I actually just created this cluster to test it. There are 6 work nodes. MongoDB only uses 3 nodes and 5 instances on each nodes ( config server, mongos, 3 replicas).  I think the resource should be enough, since there are still a lot of memory left. (6G per node) 

The issue might not be in the Chart, but I think it is good to have a discussion here. Because eventually people want to use this chart for production not just for testing. We probably need to have some special settings for this chart/kubernetes setup to make it work for production. 

Currently, I think the issue is from mongos. I am still testing it and trying to identity the issue. I will report if I find anything new.

**Comment by javsalgar** [comment]
Hi,

Agree and appreciate bringing up the discussion here. My point was that the information from the MongoDB devs is also important in order to shed some light in the cause of the issue and what we can do to fix it. Please let us know any findings. I will forward this information to our engineering team as well.

**Comment by franklin432** [comment]
how do you set the number of concurrent connections in mongo? what parameter or flag do you change to set this number @maxisam

**Comment by maxisam** [comment]
I don't think I set anything extra beyond this chart's default settings. And I think the issue is on Mongos.

**Comment by maxisam** [comment]
so I created a test using C# https://github.com/maxisam/MongoDbConnectionTest It will create 7000ish connections and you can see the errors.  (I tested it with java drive and it fails around 7000 connection as well)

**Comment by maxisam** [comment]
https://github.com/maxisam/MongoDBTest/tree/master 
I created a C# unit test project to test connection. Basically by changing the thread count, it can create more connections. 
Base on my test, it still throws exception around 7000 connections. I tested it with a java drive and the result is the same. So I think it is more on server setup.

Another thing is, using this setup, all 7000 connections are created in one mongos server.

**Comment by franklin432** [comment]
@maxisam I have a user who will be requiring 10,000 concurrent connections using the bitnami/mongodb chart so I am interested in your post. I was doing some reading and I think you were right in setting the ulimit -n ( which is the number of open files).
Mongo Docs )https://docs.mongodb.com/manual/reference/ulimit/) recommends the following at least:
-f (file size): unlimited
-t (cpu time): unlimited
-v (virtual memory): unlimited [1]
-l (locked-in-memory size): unlimited
-n (open files): 64000
-m (memory size): unlimited [1] [2]
-u (processes/threads): 64000

In addition to setting the ulimits as you did, I would think the mongo could support your 7000 connections since by default the 
net.maxIncomingConnections is set to 65536 according to (https://docs.mongodb.com/manual/reference/configuration-options/index.html). Also I'm not sure if you are using AWS resources for the mongodb-sharded chart. If so maybe you're EC2 instances need to be sized up. 
In your mongos config are you including the maxIncomingConnections parameters as such:
net:
  maxIncomingConnections: <valueHereSay65536?>

Check out this blog on someone's solution to allowing mongo to accept thousands of connections if you havent already seen it.
https://www.mongodb.com/blog/post/tuning-mongodb--linux-to-allow-for-tens-of-thousands-connections

**Comment by maxisam** [comment]
@franklin432 Thanks for your suggestion. I will go through the article you mentioned.

I know MongoDB should be able to support over 7000. But what I am not sure is MongoS. And in my test, this setup doesn't even work properly with 5000 threads. It will throw a different error but seems like not a connection issue. I think it only can support 2000 threads trying  (I haven't found the exact number). 

I didn't set maxIncomingConnections because the default is 65536 which is way higher than 7000.

My environment is on-prem VMware. Each node has 4CPU and 6G of memory. I feel like it should be enough base on the resource consumption monitor, but I will double check this part.

**Comment by dani8art** [comment]
Hi, it is just an idea, could it be caused by some of the proxies or artifact used to expose the pods to outside the cluster? how are you exposing the service?

Could you make a test just deploying it using docker and port mapping? you could use this docker-compose file making some modifications https://github.com/bitnami/bitnami-docker-mongodb-sharded/blob/master/docker-compose-multiple-shards.yml

**Comment by maxisam** [comment]
@dani8art no, I don't think so. I got error messages from mongos. Actually I am about to test that way too. I was post after.

**Comment by maxisam** [comment]
So I tested it with WSL2 (ubuntu 20) by using 4.4 from https://github.com/bitnami/bitnami-docker-mongodb-sharded. Everything works. 
ulimit is the default value

-t: cpu time (seconds)              unlimited
-f: file size (blocks)              unlimited
-d: data seg size (kbytes)          unlimited
-s: stack size (kbytes)             8192
-c: core file size (blocks)         0
-m: resident set size (kbytes)      unlimited
-u: processes                       102322
-n: file descriptors                1024
-l: locked-in-memory size (kbytes)  64
-v: address space (kbytes)          unlimited
-x: file locks                      unlimited
-i: pending signals                 102322
-q: bytes in POSIX msg queues       819200
-e: max nice                        0
-r: max rt priority                 0
-N 15:                              unlimited

With K8s+this chart, there were connection issue and some other unknown issues I haven't figured out. In the end, it always throws exceptions.
So something is wrong with my k8s setting or this chart. 

My k8s is pretty much using the default settings from Rancher 2.5.2.
For storage, I use OpenEBS local hostpath, which has near disk performance. 

I hope someone can test this

**Comment by javsalgar** [comment]
Thanks for sharing your feedback. Let's see if someone else can provide more information about their k8s environment.

**Comment by maxisam** [comment]
I have been testing this. Currently I use YCSB to test the cluster.

 It is kind of an official benchmark solution. (https://www.mongodb.com/blog/post/performance-best-practices-benchmarking)
``` yml
version: '3.5'
services:
  ycsb-stresser:
    container_name: k8s_mongo_ycsb_test
    image: stutzlab/ycsb-stresser
    environment:
      - MONGO_URL=mongodb://192.168.225.110:27017/admin?w=0
      - WORKLOAD_TYPE=a
      - RECORD_COUNT=500000
      - OPERATION_COUNT=500000
      - LOAD_THREADS=16
      - RUN_THREADS=4

```
For some reason I always get warning message like this with this chart.
```
19:22:44.324 [Thread-16] WARN  org.mongodb.driver.connection - Got socket exception on connection [connectionId{localValue:4}] to 192.168.225.110:27017. All connections to 192.168.225.110:27017 will be closed.
```

But I don't get anything like this when test against normal mongodb setup.

The performance is like half of the normal setup. However, i don't think it really uses its full potential judging by the CPU/Memory usage.

**Comment by dani8art** [comment]
> I think I found the cause of the waring message. One of the pod seems like using a lot of memory. after I set resource restriction. The problem goes away.

Does the performance go to an acceptable rate too?

**Comment by maxisam** [comment]
@dani8art Here is the test result. The testing was not on a sharding database. (I don't know how to make ycsb to work with sharding)

test setting:

      - WORKLOAD_TYPE=a
      - RECORD_COUNT=800000
      - OPERATION_COUNT=800000
      - LOAD_THREADS=16
      - RUN_THREADS=4

result:

[OVERALL], RunTime(ms), 2262430
[OVERALL], Throughput(ops/sec), 3536.0210039647636
[TOTAL_GCS_Copy], Count, 4133
[TOTAL_GC_TIME_Copy], Time(ms), 3256
[TOTAL_GC_TIME_%_Copy], Time(%), 0.14391605486136588
[TOTAL_GCS_MarkSweepCompact], Count, 0
[TOTAL_GC_TIME_MarkSweepCompact], Time(ms), 0
[TOTAL_GC_TIME_%_MarkSweepCompact], Time(%), 0.0
[TOTAL_GCs], Count, 4133
[TOTAL_GC_TIME], Time(ms), 3256
[TOTAL_GC_TIME_%], Time(%), 0.14391605486136588
[CLEANUP], Operations, 16
[CLEANUP], AverageLatency(us), 14637.3125
[CLEANUP], MinLatency(us), 1
[CLEANUP], MaxLatency(us), 234239
[CLEANUP], 95thPercentileLatency(us), 5
[CLEANUP], 99thPercentileLatency(us), 234239
[INSERT], Operations, 8000000
[INSERT], AverageLatency(us), 4508.52449225
[INSERT], MinLatency(us), 13
[INSERT], MaxLatency(us), 6414335
[INSERT], 95thPercentileLatency(us), 52
[INSERT], 99thPercentileLatency(us), 1914
[INSERT], Return=OK, 8000000


[OVERALL], RunTime(ms), 2507691
[OVERALL], Throughput(ops/sec), 3190.1857126735313
[TOTAL_GCS_Copy], Count, 6971
[TOTAL_GC_TIME_Copy], Time(ms), 4659
[TOTAL_GC_TIME_%_Copy], Time(%), 0.18578844044182477
[TOTAL_GCS_MarkSweepCompact], Count, 1
[TOTAL_GC_TIME_MarkSweepCompact], Time(ms), 21
[TOTAL_GC_TIME_%_MarkSweepCompact], Time(%), 8.37423749576802E-4
[TOTAL_GCs], Count, 6972
[TOTAL_GC_TIME], Time(ms), 4680
[TOTAL_GC_TIME_%], Time(%), 0.18662586419140156
[READ], Operations, 3999363
[READ], AverageLatency(us), 4946.638031606533
[READ], MinLatency(us), 568
[READ], MaxLatency(us), 1367039
[READ], 95thPercentileLatency(us), 12727
[READ], 99thPercentileLatency(us), 24047
[READ], Return=OK, 3999363
[CLEANUP], Operations, 8
[CLEANUP], AverageLatency(us), 677.375
[CLEANUP], MinLatency(us), 1
[CLEANUP], MaxLatency(us), 5407
[CLEANUP], 95thPercentileLatency(us), 5407
[CLEANUP], 99thPercentileLatency(us), 5407
[UPDATE], Operations, 4000637
[UPDATE], AverageLatency(us), 58.15899542997778
[UPDATE], MinLatency(us), 12
[UPDATE], MaxLatency(us), 757247
[UPDATE], 95thPercentileLatency(us), 92
[UPDATE], 99thPercentileLatency(us), 202
[UPDATE], Return=OK, 4000637


----
env  setting: 

wiredTigerCacheSizeGB=0.5

    resources:
      requests:
        cpu: "0.5"
        memory: "500Mi"
      limits:
        cpu: "0.8"
        memory: "800Mi"

**Comment by franklin432** [comment]
so instead of using the default resources you had to manually set your own for the Mongo Pods, to not have the connection drops when you are creating over 7K connections. 

default:
```
## MongoDB containers' resource requests and limits.
## ref: http://kubernetes.io/docs/user-guide/compute-resources/
##
resources:
  # We usually recommend not to specify default resources and to leave this as a conscious
  # choice for the user. This also increases chances charts run on environments with little
  # resources, such as Minikube. If you do want to specify resources, uncomment the following
  # lines, adjust them as necessary, and remove the curly braces after 'resources:'.
  limits: {}
  #   cpu: 100m
  #   memory: 128Mi
  requests: {}
  #   cpu: 100m
  #   memory: 128Mi
```

Your change: 
```
resources:
  requests:
    cpu: "0.5"
    memory: "500Mi"
  limits:
    cpu: "0.8"
    memory: "800Mi"
```

**Comment by maxisam** [comment]
@franklin432 i don't think the change can fix the 7000ish connections. The change fixes other issues I have when I use ycsb to test it.

**Comment by franklin432** [comment]
@maxisam ahh ok I see.

**Comment by dani8art** [comment]
> Could you make a test just deploying it using docker and port mapping? you could use this docker-compose file making some modifications https://github.com/bitnami/bitnami-docker-mongodb-sharded/blob/master/docker-compose-multiple-shards.yml

Hi @maxisam, sorry, but I didn't catch it, did you have a chance to test this? Just to discard k8s and focus on the container performance.

**Comment by maxisam** [comment]
@dani8art 
I did. It has the same issue. I think I found something. It might be because the node crash when the volume is large, so I am trying to set resource limitation and ensure replicas are not in the same node with the primary. I am currently blocking by https://github.com/bitnami/charts/issues/4918.

**Comment by maxisam** [comment]
After adjusting resource, I got new errors like networkinterfaceexceededtimelimit on mongos and "Couldn't get a connection within the time limit" exception from the client, which makes it really hard to get pass 5000 connections

**Comment by dani8art** [comment]
> It might be because the node crash when the volume is large, so I am trying to set resource limitation and ensure replicas are not in the same node with the primary. I am currently blocking by #4918.

Regarding this, I think it is a k8s resources configuration issue, right? if I'm not wrong it would be fixed by finding a high availability configuration for the chart, wouldn't it?

**Comment by maxisam** [comment]
@dani8art yeah, I did. So I set a higher resource limit, I got a higher connection count. It is still 7000ish but 400 more. I think I can safely assume the problem is on resource, especially on memory. 

There are still some problems.

Like I still get lots of exceptions from the code when the volume go up. 
The overall performance is not as good as mongodb without container structure. (~30% lower)

However, I think this chart is not the reason for above issues. I hope someone else can do some tests on this and see if we can figure out how to optimize it better than what I have. 

And I don't know what is the correct way to adjust ulimit for this chart from initContainer

I tried to modify /etc/security/limits.conf for user 1001 but it doesn't work

in case anyone is interested my vm setup is
CPU: 4 core (2.7kHz) 
Memory: 6G 
OS: ubuntu 20.04

**Comment by dani8art** [comment]
> The overall performance is not as good as mongodb without container structure. (~30% lower)

Hi @maxisam, yes I think so.

We have created an internal task in order to perform further investigations, unfortunately, we can't give you an ETA but we will keep you updated with our investigation results.

BTW, we would appreciate it if you provide the exact instructions to reproduce your environment, helm commands, values.yaml, test scripts (if possible) and so on that permit us to continue the investigation with a starting point.

Thanks!

**Comment by maxisam** [comment]
So I use Rancher 2.5.5 to create my k8s. (1.19.4)
It has 9 nodes. (6 works nodes). 
Node is created by VMWare vSphere 6.5 
CPU: 4 core (2.7kHz)
Memory: 6G
OS: ubuntu 20.04
Storage: OpenEBS HostPath
Network: Weave
LoadBalancer: MetalLB
DNS: CoreDNS
Helm: 3

values.yml
```yml
## Global Docker image parameters
## Please, note that this will override the image parameters, including dependencies, configured to use the global value
## Current available global Docker image parameters: imageRegistry, imagePullSecrets and storageClass
##
global:
  #   imageRegistry: myRegistryName
  #   imagePullSecrets:
  #     - myRegistryKeySecretName
  storageClass: "openebs-hostpath"

## Bitnami MongoDB Sharded image version
## ref: https://hub.docker.com/r/bitnami/mongodb-sharded/tags/
##
image:
  registry: docker.io
  repository: bitnami/mongodb-sharded
  tag: 4.4.3-debian-10-r21
  ## Specify a imagePullPolicy
  ## Defaults to 'Always' if image tag is 'latest', else set to 'IfNotPresent'
  ## ref: http://kubernetes.io/docs/user-guide/images/#pre-pulling-images
  ##
  pullPolicy: IfNotPresent
  ## Optionally specify an array of imagePullSecrets.
  ## Secrets must be manually created in the namespace.
  ## ref: https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/
  ##
  pullSecrets:

  ## Set to true if you would like to see extra information on logs
  ## It turns on Bitnami debugging in minideb-extras-base
  ## ref:  https://github.com/bitnami/minideb-extras-base
  ##
  debug: false

## MongoDB credentials
##

## MongoDB root password
## If set to null it will be randomly generated
## ref: https://github.com/bitnami/bitnami-docker-mongodb/blob/master/README.md#setting-the-root-password-on-first-run
##
## mongodbRootPassword: password
##

## Create a non-root MongoDB user
## ref: https://github.com/bitnami/bitnami-docker-mongodb/blob/master/README.md#setting-the-root-password-on-first-run
##
## mongodbUsername: user
##
mongodbUsername: admin

## non-root MongoDB user password
## ref: https://github.com/bitnami/bitnami-docker-mongodb/blob/master/README.md#setting-the-root-password-on-first-run
##
## mongodbPassword: password
##
## Create database on first run
## ref: https://github.com/bitnami/bitnami-docker-mongodb/blob/master/README.md#setting-the-root-password-on-first-run
##
# mongodbDatabase: database
mongodbDatabase: test

## Replica Set key (shared for shards and config servers)
## ref:
##
## replicaSetKey: testkey123
##

## Name of a secret containing all the credentials above
## ref:
##
## existingSecret: name-of-existing-secret
##
existingSecret: mongodb-secret

## Mount credentials as files instead of using environment variables
##
usePasswordFile: true

## Number of MongoDB Shards
## ref: https://docs.mongodb.com/manual/core/sharded-cluster-shards/
##
shards: 4

## Shard replica set properties
## ref: https://docs.mongodb.com/manual/replication/index.html
##
shardsvr:
  ## Properties for data nodes (primary and secondary)
  ##
  dataNode:
    ## Number of replicas. A value of replicas=1 is simply a primary node
    ##
    replicas: 3
    ## Configure resource requests and limits
    ## ref: http://kubernetes.io/docs/user-guide/compute-resources/
    ##
    resources:
      requests:
        cpu: "0.5"
        memory: "500Mi"
      limits:
        cpu: "0.8"
        memory: "800Mi"
    ## MongoDB additional command line flags
    ##
    ## Can be used to specify command line flags, for example:
    ##
    ## mongodbExtraFlags:
    ##  - "--wiredTigerCacheSizeGB=2"
    ##
    mongodbExtraFlags:
      - "--wiredTigerCacheSizeGB=0.5"
    ## Pod priority
    ## https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/
    ##
    priorityClassName:
    ## Node selector
    ## ref: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#nodeselector
    ##
    nodeSelector:
      beta.kubernetes.io/os: linux
    ## Affinity
    ## ref: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#affinity-and-anti-affinity
    ##
    affinity:
      podAntiAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchLabels:
                shard: "{{ .dataNodeLoopId }}"
            topologyKey: kubernetes.io/hostname

    ## Tolerations
    ## ref: https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/
    ##

    serviceAccount:
      ## Specifies whether a ServiceAccount should be created.
      ##
      create: true
      ## The name of the ServiceAccount to use.
      ## If not set and create is true, a name is generated using the XXX.fullname template
      ##
      name: mongo-shard-data

  ## Properties for arbiter nodes
  ## ref: https://docs.mongodb.com/manual/tutorial/add-replica-set-arbiter/
  ##
 
  persistence:
    enabled: true
    ## The path the volume will be mounted at, useful when using different
    ## MongoDB images.
    ##
    mountPath: /bitnami/mongodb

    ## The subdirectory of the volume to mount to, useful in dev environments
    ## and one PV for multiple services.
    ##
    subPath: ""

    ## mongodb data Persistent Volume Storage Class
    ## If defined, storageClassName: <storageClass>
    ## If set to "-", storageClassName: "", which disables dynamic provisioning
    ## If undefined (the default) or set to null, no storageClassName spec is
    ##   set, choosing the default provisioner.  (gp2 on AWS, standard on
    ##   GKE, AWS & OpenStack)
    ##
    # storageClass: "-"
    accessModes:
      - ReadWriteOnce
    ## PersistentVolumeClaim size
    ##
    size: 10Gi
    ## Additional volume annotations
    ## ref: https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/
    ##
    annotations: {}

## Config Server replica set properties
## ref: https://docs.mongodb.com/manual/core/sharded-cluster-config-servers/
##
configsvr:
  ## Number of replicas. A value of replicas=1 is simply a primary node
  ##
  replicas: 3
  ## Configure resource requests and limits
  ## ref: http://kubernetes.io/docs/user-guide/compute-resources/
  ##
  resources:
    requests:
      memory: "200Mi"
  
  affinity:
   podAntiAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
        - podAffinityTerm:
            labelSelector:
              matchLabels:
                app.kubernetes.io/component: configsvr
            topologyKey: kubernetes.io/hostname
          weight: 100

  ## Tolerations
  ## ref: https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/
  ##
  
  persistence:
    enabled: true
    ## The path the volume will be mounted at, useful when using different
    ## MongoDB images.
    ##
    mountPath: /bitnami/mongodb

    ## The subdirectory of the volume to mount to, useful in dev environments
    ## and one PV for multiple services.
    ##
    subPath: ""

    ## mongodb data Persistent Volume Storage Class
    ## If defined, storageClassName: <storageClass>
    ## If set to "-", storageClassName: "", which disables dynamic provisioning
    ## If undefined (the default) or set to null, no storageClassName spec is
    ##   set, choosing the default provisioner.  (gp2 on AWS, standard on
    ##   GKE, AWS & OpenStack)
    ##
    storageClass: "openebs-hostpath"
    accessModes:
      - ReadWriteOnce
    ## PersistentVolumeClaim size
    ##
    size: 3Gi
    ## Additional volume annotations
    ## ref: https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/
    ##
    annotations: {}
  ## K8s Service Account.
  ## ref: https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/
  ##
  serviceAccount:
    ## Specifies whether a ServiceAccount should be created.
    ##
    create: true
    ## The name of the ServiceAccount to use.
    ## If not set and create is true, a name is generated using the XXX.fullname template
    ##
    name: mongo-configsvr


## Mongos properties
## ref: https://docs.mongodb.com/manual/reference/program/mongos/#bin.mongos
##
mongos:
  ## Number of replicas
  ##
  replicas: 3
  ## Configure resource requests and limits
  ## ref: http://kubernetes.io/docs/user-guide/compute-resources/
  ##
  resources:
    requests:
      cpu: "2"
      memory: "3000Mi"
    limits:
      cpu: "3"
      memory: "4000Mi"
  ## MongoDB additional command line flags
  ##
  ## Can be used to specify command line flags, for example:
  ##
  mongodbExtraFlags:
    - --maxConns=100000

  affinity:
    podAntiAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
        - podAffinityTerm:
            labelSelector:
              matchExpressions:
                - key: app.kubernetes.io/name
                  operator: In
                  values:
                    - mongos
            topologyKey: kubernetes.io/hostname
          weight: 100
  ## Tolerations
  ## ref: https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/
  ##


## Properties for all of the pods in the cluster (shards, config servers and mongos)
##
common:
  ## Maximum peer node waiting timeout (in seconds)
  ##
  mongodbMaxWaitTimeout: 120
  initContainers:
    - name: "env-setup0"
      image: "busybox:latest"
      securityContext:
        privileged: true
        allowPrivilegeEscalation: true
      command: ["/bin/sh", "-c"]
      args:
        - >-
          id -u &&
          sysctl -w net.ipv4.tcp_keepalive_time=120 &&
          sysctl -w vm.max_map_count=262144

## Init containers parameters:
## volumePermissions: Change the owner and group of the persistent volume mountpoint to runAsUser:fsGroup values from the securityContext section.
##
volumePermissions:
  enabled: false
  image:
    registry: docker.io
    repository: bitnami/minideb
    tag: buster
    pullPolicy: Always
    ## Optionally specify an array of imagePullSecrets.
    ## Secrets must be manually created in the namespace.
    ## ref: https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/
    ##
    # pullSecrets:
    #   - myRegistryKeySecretName
  resources: {}

## Pod Security Context
## ref: https://kubernetes.io/docs/tasks/configure-pod-container/security-context/
##
securityContext:
  enabled: true
  fsGroup: 1001
  runAsUser: 1001
  runAsNonRoot: true

## Kubernetes Cluster Domain
## ref: https://kubernetes.io/docs/tasks/administer-cluster/dns-custom-nameservers/#introduction
##
clusterDomain: cluster.local

## Kubernetes service type
## ref: https://kubernetes.io/docs/concepts/services-networking/service/
##
service:
  type: LoadBalancer

  ## External traffic policy
  ## ref: https://kubernetes.io/docs/concepts/services-networking/service/#publishing-services-service-types
  ##
  externalTrafficPolicy: Cluster
  ## MongoDB Service port and Container Port
  ##
  port: 27017
  ## Set a fixed service ClusterIP addrress
  ## ref: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#servicespec-v1-core
  ##
 
  loadBalancerIP: 192.168.225.110

 
  sessionAffinity: ClientIP

#
## Prometheus Exporter / Metrics
##
metrics:
  enabled: true

  image:
    registry: docker.io
    repository: bitnami/mongodb-exporter
    tag: 0.20.1-debian-10-r37
    pullPolicy: Always

  ## Metrics exporter pod Annotation
  ##
  podAnnotations:
    prometheus.io/scrape: "true"
    prometheus.io/port: "9216"

  # Enable this if you're using https://github.com/coreos/prometheus-operator
  serviceMonitor:
    enabled: true
    namespace: cattle-monitoring-system
  ## Prometheus Service Monitor
  ## ref: https://github.com/coreos/prometheus-operator
  ##      https://github.com/coreos/prometheus-operator/blob/master/Documentation/api.md#endpoint
  ##
  podMonitor:
    ## If the operator is installed in your cluster, set to true to create a PodMonitor entry
    ##
    enabled: true
    ## Specify the namespace in which the podMonitor resource will be created
    ##
    namespace: cattle-monitoring-system

```

ycsb test (I only use it to test none-sharding collection) But you can still see the performance difference.

```yml
version: '3.5'
services:
  ycsb-stresser:
    container_name: k8s_mongo_ycsb_test
    image: stutzlab/ycsb-stresser
    environment:
      - MONGO_URL=mongodb://root:password123@mongodb-mongodb-sharded:27017/admin?w=0
      - WORKLOAD_TYPE=a
      - RECORD_COUNT=800000
      - OPERATION_COUNT=800000
      - LOAD_THREADS=16
      - RUN_THREADS=4
```

I think the best way to do this is test one mongodb without sharding to see the difference between standalone mongodb and docker.

I will keep testing, but no ETA as well.

**Comment by dani8art** [comment]
Thanks for the info!

**Comment by rafariossaa** [comment]
Hi,
  One quick question,  when connecting to the mongo cluster and found this limitation, Were you connecting to **from** other pod inside the cluster or from outside, eg., your laptop ?.

**Comment by maxisam** [comment]
@rabbituser01 I tested both.

**Comment by rafariossaa** [comment]
Hi,
 Thanks for the info.

**Comment by rafariossaa** [comment]
Hi
  I did some testing on this, but I could not find issues related to the docker images, what I did find was that the memory usage by mongo increase quite a lot with the number of connections and even I got the pod killed. 
  My guess is that this is related on the inner of kubernetes on how it handles the networking, but I could not find parameter related to this.

**TL;DR;**
  I did jmeter test similar to [this one](https://www.blazemeter.com/blog/mongodb-performance-testing-with-jmeter), I only used the connection section of the code.
```
//
//  Based on code from: https://www.blazemeter.com/blog/mongodb-performance-testing-with-jmeter
//

import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoClient;
import com.mongodb.MongoClientSettings;
import com.mongodb.ServerAddress;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.MongoCredential;

import org.bson.Document;

import java.util.Arrays;

try {
        MongoCredential credential = MongoCredential.createCredential(vars.get(&quot;mongoUser&quot;), vars.get(&quot;databaseName&quot;), vars.get(&quot;mongoPwd&quot;).toCharArray());

        MongoClientSettings settings = MongoClientSettings.builder()
             .credential(credential)
                .applyToClusterSettings {builder -&gt; 
                        builder.hosts(Arrays.asList(new ServerAddress(vars.get(&quot;mongoHost&quot;),vars.get(&quot;mongoPort&quot;).toInteger())))}
                .build();

        MongoClient mongoClient = MongoClients.create(settings);

//      MongoDatabase database = mongoClient.getDatabase(vars.get(&quot;databaseName&quot;));
//      MongoCollection&lt;Document&gt; collection = database.getCollection(vars.get(&quot;collectionName&quot;));

//     vars.putObject(&quot;collection&quot;, collection);

     // milliseconds a connection will be opend
     Thread.sleep(10000)

     mongoClient.close();

        return &quot;Connected to &quot; + vars.get(&quot;collectionName&quot;);
}
catch (Exception e) {
        SampleResult.setSuccessful(false);
        SampleResult.setResponseCode(&quot;500&quot;);
        SampleResult.setResponseMessage(&quot;Exception: &quot; + e);
}
```
  I used 1,10,100,1000,10000 threads and set three different scenarios, docker, minikube and a AKS cluster, where I used the provided docker-compose.yaml, and the default chart parameters. I only set the user password to be able to configure the connection parameters in jmeter. I checkeck the connection number by the output of the test and also in mongodb by running in a terminal:
```
watch -n 1 mongo  -u root -p $MONGODB_ROOT_PASSWORD --eval "'db.serverStatus().connections'"
```

  For the kubernetes tests, I deployed the jmeter, outside the cluster, inside de cluster as a different pod and as other container is the same pod. As it seems network related, the idea behind is use less and less layers of networking in each test scenario.

  With docker and minikube, I found not issues. I noticed that in minikube the connection rate was slower than with just docker, but I could reach 8000 connections without issues.
  In AKS, was the same, the connection rate was a bit more than slower than in minkube but I could reach 8000 connections. The only exception was when I deployed jmeter inside the same pod that I got the pod killed because of the ram usage.

  I could also notice in all the tests scenarios, that mongodb uses quite a lot of RAM when handling the connections, and I was only opening and closing the connections after 10seconds, there was not other operation involved. In the [mongodb documentation](https://docs.mongodb.com/manual/reference/ulimit/#resource-utilization) I cloud find that the connections uses a file descriptor and a new thread, and this reminds me of the [C10k problem ](https://en.wikipedia.org/wiki/C10k_problem).
   By this last, and also how connection rate is slower in kubernetes, I guess that what is happening here is that the usage of resources from mongo is hitting throttling or a limit by the underlying system, it could be the operating system or K8s or both. I could not find exactly if there is a parameter that control this.

**Comment by maxisam** [comment]
@rafariossaa Thanks for testing. I did that JMeter test as well and reference to the same website ha. Yeah, your finding is aligned with my result. Somehow it uses a tons of memory when connections get to a point. I doubled the memory but only got 400ish more wiggle room. Could you share how many memory you assigned for one instance? did you test one standalone mongodb or sharding setup? thx!

**Comment by rafariossaa** [comment]
Hi,
  I used the default `values.yml` from the chart, and installed it with:
```
helm install mymongo bitnami/mongodb --set auth.rootPassword=mypassword
```
  I haven't tried the mongodb sharded, as I am using only `open connection` I am supposing that it should not make a difference. I should be different in the case of handling data.

**Comment by carrodher** [comment]
Unfortunately, this issue was created a long time ago and although there is an internal task to fix it, it was not prioritized as something to address in the short/mid term. It's not a technical reason but something related to the capacity since we're a small team.

Being said that, contributions via PRs are more than welcome in both repositories (containers and charts). Just in case you would like to contribute.

During this time, there are several releases of this asset and it's possible the issue has gone as part of other changes. If that's not the case and you are still experiencing this issue, please feel free to reopen it and we will re-evaluate it.
